pennylane/devices/default_qubit_torch.py

```diff
@@ -0,0 +1,301 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains a PyTorch implementation of the :class:`~.DefaultQubit`
+reference plugin.
+"""
+
+try:
+    import torch
+
+    v = torch.__version__.split(".")
+    if int(v[1]) < 8 or (int(v[1]) == 8 and int(v[2][0]) == 0):
+        raise ImportError("default.qubit.torch device requires Torch>=1.8.0")
+
+except ImportError as e:
+    raise ImportError("default.qubit.torch device requires Torch>=1.8.0") from e
+
+import warnings
+from string import ascii_letters as ABC
+import numpy as np
+from pennylane.operation import DiagonalOperation
+from pennylane.devices import torch_ops
+from . import DefaultQubit
+
+ABC_ARRAY = np.array(list(ABC))
+
+
+class DefaultQubitTorch(DefaultQubit):
+    """Simulator plugin based on ``"default.qubit"``, written using PyTorch.
+
+    **Short name:** ``default.qubit.torch``
+
+    This device provides a pure-state qubit simulator written using PyTorch.
+    As a result, it supports classical backpropagation as a means to compute the Jacobian. This can
+    be faster than the parameter-shift rule for analytic quantum gradients
+    when the number of parameters to be optimized is large.
+
+    To use this device, you will need to install PyTorch:
+
+    .. code-block:: console
+
+        pip install torch>=1.8.0
+
+    **Example**
+
+    The ``default.qubit.torch`` is designed to be used with end-to-end classical backpropagation
+    (``diff_method="backprop"``) and the PyTorch interface. This is the default method
+    of differentiation when creating a QNode with this device.
+
+    Using this method, the created QNode is a 'white-box', and is
+    tightly integrated with your PyTorch computation:
+
+    .. code-block:: python
+
+     dev = qml.device("default.qubit.torch", wires=1)
+     @qml.qnode(dev, interface="torch", diff_method="backprop")
+     def circuit(x):
+         qml.RX(x[1], wires=0)
+         qml.Rot(x[0], x[1], x[2], wires=0)
+         return qml.expval(qml.PauliZ(0))
+     weights = torch.tensor([0.2, 0.5, 0.1], requires_grad=True)
+     res = circuit(weights)
+     res.backward()
+    >>> print(weights.grad)
+     tensor([-2.2527e-01, -1.0086e+00,  1.3878e-17])
+
+    Autograd mode will also work when using classical backpropagation:
+
+    >>> def cost(weights):
+    ...    return torch.sum(circuit(weights)**3) - 1
+    >>> res = circuit(weights)
+    >>> res.backward()
+    >>> print(weights.grad)
+    tensor([-3.5472e-01, -1.5883e+00,  2.0817e-17])
+
+    Executing the pipeline in PyTorch will allow the whole computation to be run on the GPU,
+    and therefore providing an acceleration. Your parameters need to be instantiated on the same
+    device as the backend device.
+
+    .. code-block:: python
+
+     dev = qml.device("default.qubit.torch", wires=1, torch_device='cuda')
```

Comment:
Since our tests can't cover `torch_device='cuda'`, and I don't personally have a GPU to test this on, I'd like to double-check that this in fact works.

Reply:
Have tested this locally, but run into an issue all the way down in `default_qubit`'s `_apply_unitary_einsum` method.

```
RuntimeError: Expected tensor to have cuda DeviceType, but got tensor with cpu DeviceType (while checking arguments for einsum())
```

I have copied the example directly from the documentation; @Slimane33 have you run into similar issues? Is there any additional configuration needed that is not explicitly shown in the example?

Reply:
Yes, after some investigation, the problem comes from the Qnode construction itself, at the beginning of the method `execute` the `self._state` is on the CPU when it should be on the GPU. I suggest to do something like 
```python
if hasattr(self, '_torch_device'):
     self._state = self.state.to(self._torch_device)
```

Reply:
It is fixed following my method: see  https://colab.research.google.com/drive/1Xb_-l3TIOZhbDw6K9jO34oDgDqzs3o7V?usp=sharing

Reply:
@josh146 @Slimane33 Just want to know which part we need to fix, got bit confused in the discussions part.

Reply:
The initial error appears to be fixed with the most recent updates (just involved making sure that all the necessary data is on the GPU); I am just not able to fully test now as I run into further issues, but they are from pytorch itself, and may be partially related to the type of graphics card I have. I will try again later today to see if I can get things up and running.

Reply:
I was able to get the example to work, but I needed to use the development version (v. 1.9) of PyTorch. I was experiencing the same issue documented [here](https://github.com/allenai/allennlp/issues/5064), which is fixed in 1.9. As a consequence I'm not able to run the device test suite, since it requires v. 1.8. Trying with 1.9 causes errors to occur in three `TestPassthruIntegration` tests due to conversion of numpy arrays. I'm not sure if that's due to  differences in 1.9, or issues with the functions or tests themselves.

Reply:
@glassnotes  Just for confirmation, ``` TestPassthruIntegration ``` are coming when you run using GPU device or CPU?

Reply:
I believe CPU, since the test file doesn't appear to be using the GPU device. Specifically,

```
>  grad_matrix = torch.as_tensor(torch.from_numpy(grad_matrix), dtype=tape.dtype)
E TypeError: can't convert np.ndarray of type numpy.object_. The only supported types are: float64, float32, float16, complex64, complex128, int64, int32, int16, int8, uint8, and bool.
```
This comes from the  `_evaluate_grad_matrix` function. It is unhappy about how the conversion between `torch` and `numpy` arrays is happening, but again, this could be due to differences between versions of `torch` (especially if the test suite is passing on the CI).

Reply:
Hey @Slimane33 and @arshpreetsingh just wanted to give you an update, we are getting some more people on our team to test the GPU version locally. In addition to the issues with running on the GPU testing, we are getting `ImportError`s with `torch` due to string parsing in the version check, could you please update this to use `semantic_version` as is done in some of the other files?

Reply:
Hi @glassnotes thanks for the update I will look at it.